### ISSUE TITLE
Delete unsaved uploaded files

### ISSUE THREAD

**Issue by uzi0espil** [bhch/react-json-form#38]
Since the upload of the file is done through Ajax call to an endpoint specified in Django `settings.JSONFORM_UPLOAD_HANDLER`, then there might be a case where the client would upload and close the page or cancel saving the model and the files would be still be uploaded without any link to a record in the database. In this case, how would you remove these uploaded files.

**Proposal:**

One way I can think of, is when a file is successfully uploaded, then in the `.then((result)` we store the returned file name in a list in react, and upon exit of page without saving (like explained [here](https://stackoverflow.com/questions/32841757/detecting-user-leaving-page-with-react-router)), if the list is not empty, then we send a DELETE to `settings.JSONFORM_UPLOAD_HANDLER` to delete those files. The clear button should as well send DELETE.

In this case, the handler endpoint should deal with both POST and DELETE methods.

**Comment by bhch** [comment]
> files would be still be uploaded without any link to a record in the database

During the upload process, your handler function is called and there you can save the file in a dedicated media model. That way you'll get a record in the database.

---

Anyway, I like the general idea you're proposing. I will mark this issue as "planned for future". I'll have to think about it a bit more to start working on it.

Thank you for the suggestion.

**Comment by uzi0espil** [comment]
Give it a thought, I am willing to contribute to the project. If agreed on the method.

**Comment by bhch** [comment]
Right now I'm rewriting some internal parts of the widget so that it can also be used as a library in custom JS projects outside of Django. Let's wait a few days until after I get this version out.

**Comment by bhch** [comment]
In the new version, I've implemented a *"Choose from library"* feature ([see screenshot here](https://django-jsonform.readthedocs.io/en/stable/_images/file-upload.gif)).

Now a modal will open up and you have the ability to display previously uploaded files to the user.

I think we can display a "Delete" button under the thumbnails so user can manually delete the files if they want. To me it seems better than automatic deletion.

[Read the docs](https://django-jsonform.readthedocs.io/en/stable/guide/upload.html#file-url) on how to implement this feature and let me know your feedback.

**Comment by uzi0espil** [comment]
Sorry for the late delay, I just got the chance to try it out the new version 2.11.1

To be honest, the changes you did doesn't suit my use case. In my case, I am not uploading images, I am uploading parquet files (huge files that can reach 500MB of size each file), I don't want to the client to look at the what he uploaded previously and I want one single parquet file per new jsonform. So if I have created 5 models that uses JSONForm, then I want only 5 files. If during filling the form upload has made but for some reason the client didn't commit the changes, then the file should be removed.

I would like to thank you for your effor but unfortunately the changes you made, made the library unusable for my case as you are forcing the user to GET all the files and select from one to assign a file to the variable.

**Comment by bhch** [comment]
>  made the library unusable for my case as you are forcing the user to GET all the files and select from one to assign a file to the variable.

There's no forcing of any sort. The upload feature still works as it did earlier. The user can still upload new files.

The *"Choose from library"* feature is left completely up to you to implement. You can return empty results for GET requests if you want.

---

Anyway, I will provide support for deleting files. A DELETE request will be sent when:

 - page is closed without saving
 - the "Clear" button is clicked
 - the "Delete" button is clicked (this button will be shown under thumbnails of previously uploaded files)

Also, each of these events will send an `event` parameter with the DELETE request so that we can identify on the server what triggered the deletion.

For example: `DELETE /path/to/file.jpg?event=page_exit`.

So, it is up to you whether you want to delete the file on page exit or keep it.

**Comment by uzi0espil** [comment]
Thanks alot for the quick reply.

> There's no forcing of any sort. The upload feature still works as it did earlier. The user can still upload new files.
> 
> The "Choose from library" feature is left completely up to you to implement. You can return empty results for GET requests if you want.

Yes it seems to work. First, when I tried to upload a file, it uploads correctly, and the modal window will close. However, there is no `clear` button and the path isn't displayed next to the input. When I try to save, it fails because the data-url input is empty.

The problem was that I returned `JsonResponse({'file_path': file_url})` like what was in previous versions, once I changed to `JsonResponse({'value': file_url})`, then it worked accordingly.

It would be nice though to give control to whether show or hide the `Choose from library` tab, currently I am setting the style of class `rjf-modal__header` to display:none to hide it.

<hr />

> Also, each of these events will send an event parameter with the DELETE request so that we can identify on the server what triggered the deletion.
> 
> For example: DELETE /path/to/file.jpg?event=page_exit.
> 
> So, it is up to you whether you want to delete the file on page exit or keep it.

This would be a great addition to the library!

Thanks again for the effort you put on the library!

**Comment by bhch** [comment]
So I tried some things. There is no reliable way to send AJAX requests before the user closes the page.

**1. Example without prompt:**

This doesn't work at all. No request is sent by the browser or the request is immediately terminated and never reaches the server.

```javascript
window.addEventListener('beforeunload', function(e) {
    fetch('/file-handler/', {method: 'DELETE'});
});
```

**2. Example with prompt:**

This works, but is unreliable.

 - If the request takes too long to reach server and the user confirms exit prompt before that, the browser will terminate the request.
 - If user doesn't exit the page, the request will have already been sent. So the file will be deleted.

```javascript
window.addEventListener('beforeunload', function(e) {
    e.preventDefault();
    e.returnValue = ''; // chrome fix

    fetch('/file-handler/', {method: 'DELETE'});
});
```

---

If you have any other ideas, let's discuss them.

**Comment by bhch** [comment]
Most reliable solution so far is using `fetch` with `keepalive` flag. This works on most modern browsers.

```javascript
/* ask exit confirmation */
window.addEventListener('beforeunload', function(e) {
    e.preventDefault();
    e.returnValue = ''; // chrome fix
});

/* send delete request */
window.addEventListener('unload', function(e) {
    fetch('/file-handler/', {method: 'DELETE', keepalive: true});
});
```

**Comment by bhch** [comment]
Hello. I've implemented the file deletion support in [v2.12.0](https://github.com/bhch/django-jsonform/releases/tag/v2.12.0).

Please upgrade and test it out. Also [read the docs](https://django-jsonform.readthedocs.io/en/latest/guide/upload.html#deleting-files) on this feature.

---

> It would be nice though to give control to whether show or hide the Choose from library tab

I will add something in the next release. I'm thinking of provided two file upload widgets: one a simple input (like it used to be in earlier versions) and another widget which opens the modal.

I've created a new issue to track this: #50.
